# Generating CATH assignments for large sequence datasets

## About this document

 * This document provides a set of instructions to annotate large sequence datasets with CATH functional family assignments.
 * This document is currently considered 'beta' and is subject to change.
 * All feedback / corrections / suggestions are welcome

# Steps

## 1. Create a new project

Clone a new project template from GitHub and move into the directory.

```
> git clone https://github.com/UCLOrengoGroup/cath-tools-genomescan.git myproject
> cd myproject/
> pwd
```

Note: this directory will be called ```${PROJECTHOME}``` for the rest of the tutorial.

## 2. Organise your query sequences

Choose a genome and get all translated protein sequences in FASTA format. 

The file ```${PROJECTHOME}/data/test.fasta``` has been included in the template as an example.

## 3. Download and install HMMER3

Download and extract the HMMER3 software:

```
> cd ${PROJECTHOME}
> cd apps/
> wget http://eddylab.org/software/hmmer3/3.1b2/hmmer-3.1b2-linux-intel-x86_64.tar.gz
> tar zxf hmmer-3.1b2-linux-intel-x86_64.tar.gz
```

Create a link for convenience:

```
> cd ${PROJECTHOME}
> cd bin/
> ln -s ../apps/hmmer-3.1b2-linux-intel-x86_64/binaries hmmer3
```

Check to make sure the link and binaries work as expected:

```
> cd ${PROJECTHOME}
> ./bin/hmmer3/hmmscan -h
# hmmscan :: search sequence(s) against a profile database
# HMMER 3.1b2 (February 2015); http://hmmer.org/
# Copyright (C) 2015 Howard Hughes Medical Institute.
# Freely distributed under the GNU General Public License (GPLv3).
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Usage: hmmscan [-options] <hmmdb> <seqfile>
...
```

More information can be found on installing HMMER3 on the excellent [HMMER documentation](http://hmmer.org/documentation.html)

## 4. Download library of CATH models

Download the library of CATH FunFam (Functional Family) models and unzip:

```
> cd ${PROJECTHOME}
> cd data/
> wget http://download.cathdb.info/cath/releases/all-releases/v4_1_0/sequence-data/funfam-hmm3-v4_1_0.lib.gz
> gunzip funfam-hmm3-v4_1_0.lib.gz
```

Use ```hmmpress``` to index the HMM library:

```
> cd ${PROJECTHOME}
> cd data/
> ../bin/hmmer3/hmmpress funfam-hmm3-v4_1_0.lib
```

## 5. Scan the query sequence(s)

Query sequences(s) should be in the FASTA format.
An example fasta file is available in the data/ folder.

```
> cd ${PROJECTHOME}
> cd data/
> ls test.fasta
>cath|4_1_0|2fupA
GXPDSPTLLDLFAEDIGHANQLLQLVDEEFQALERRELPVLQQLLGAKQPLXQQLERNGRARAEILREAGVSLDREGLARYARERADGAELLARGDELGELLERCQQANLRNGRIIRANQASTGSLLNILR
>cath|4_1_0|1cukA
MIGRLRGIIIEKQPPLVLIEVGGVGYEVHMPMTCFYELPEAGQEAIVFTHFVVREDAQLLYGFNNKQERTLFKELIKTNGVGPKLALAILSGMSAQQFVNAVEREEVGALVKLPGIGKKTAERLIVEMKDRFKGLHGDLFTPAADLVLTSPASPATDDAEQEAVARLVALGYKPQEASRMVSKIARPDASSETLIREALRAAL
```

CATH FunFam assignments can now be generated by scanning the query sequences against the CATH FunFam model library. Only those which meet the FunFam inclusion threshold are reported.

```
> cd ${PROJECTHOME}
> ./apps/cath-genomescan.pl -i data/test.fasta -l data/funfam-hmm3-v4_1_0.lib -o results/
```

## 6. Results

Three result files are generated in the results folder:
```
> cd ${PROJECTHOME}/results/
> ls
 test.crh  test.domtblout  test.html
```

#### (i) test.crh 
This is the main FunFam scan result file which collapses a list of domain matches to your query sequence(s) down to the best, non-overlapping subset (ie domain architecture).

```
> cd ${PROJECTHOME}/results/
> cat test.crh
# Generated by cath-resolve-hits, one of the cath-tools (https://github.com/UCLOrengoGroup/cath-tools)
#FIELDS query-id match-id score boundaries resolved cond-evalue indp-evalue
cath|4_1_0|1cukA 2.40.50.140/FF/58874 160.3 1-65 1-65 5.4e-52 1.1e-51
cath|4_1_0|1cukA 1.10.150.20/FF/4839 169.4 67-140 67-140 5.5e-55 1.1e-54
cath|4_1_0|1cukA 1.10.8.10/FF/15460 61.4 157-201 157-201 3.8e-21 7.5e-21
cath|4_1_0|2fupA 1.20.58.300/FF/1517 113 6-130 6-130 5.7e-37 1.1e-36
```

#### (ii) test.domtblout
This is the raw FunFam scan file generated by HMMER3 which lists all the domain matches of your query sequence(s) to CATH FunFam models.

#### (iii) test.html
Point your web browser at this file to visualise the assigned domain architecture of your query sequence(s).

# Relevant Links:

1. [Functional classification of CATH superfamilies: a domain-based approach for protein function annotation](https://doi.org/10.1093/bioinformatics/btv398)

2. [CATH FunFHMMer web server: protein functional annotations using functional family assignments](https://doi.org/10.1093/nar/gkv488)

3. [HMMER3 User Manual](ftp://ftp.hgc.jp/pub/mirror/wustl/hmmer3/3.1b1/Userguide.pdf)

